[GTK][WPE] Propagate the damage from accelerated 2D canvas

Scony opened the pull request:
#### fdfdd4c0e355483c04b7c4e3b4ccae00c23bfcf7
<pre>
[GTK][WPE] Propagate the damage from accelerated 2D canvas
<a href="https://bugs.webkit.org/show_bug.cgi?id=291678">https://bugs.webkit.org/show_bug.cgi?id=291678</a>

Reviewed by Nikolas Zimmermann.

This change enables propagating damage from accelerated 2D canvas up
to the platform and to the WPE/GTK WebKit&apos;s compositor so that it can
optimize some of its drawing operations.

Co-authored-by: Carlos Garcia Campos &lt;cgarcia@igalia.com&gt;
Canonical link: <a href="https://commits.webkit.org/298552@main">https://commits.webkit.org/298552@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/955baa95ba394762625a26bd77df3de151cdf315

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115737 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35400 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25923 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121785 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66270 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117626 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36081 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43983 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87917 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118685 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28783 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103858 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68318 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27924 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21973 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65456 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98171 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22094 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124935 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42632 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31969 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96674 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42999 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100047 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96460 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24563 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41720 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19578 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38557 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42523 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48095 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41996 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45327 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43704 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6c57859339c568acba409c161d879366153bef26. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44192

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/24ac243fe62beb014f79b824e79565422ef1f1b5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44192

carlosgcampos commented on Source/WebCore/rendering/RenderBoxModelObject.cpp:
```diff
@@ -167,11 +167,19 @@ void RenderBoxModelObject::setSelectionState(HighlightState state)
     if (!hasLayer())
         return;
 
+#if ENABLE(DAMAGE_TRACKING)
+    if (dirtyRect && layer()->backing() && layer()->backing()->graphicsLayer())
+        layer()->backing()->graphicsLayer()->setContentsNeedsDisplayInRect(*dirtyRect);
+#endif
```
Same here, we can add setContentsNeedsDisplayInRect unconditionally and when damage tracking is not enabled it just calls setContentsNeedsDisplay

Scony commented on Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
```diff
@@ -336,6 +336,26 @@ void GraphicsLayerCoordinated::setContentsNeedsDisplay()
+    if (!m_contentsVisible || m_size.isEmpty())
+        return;
+
+    if (m_dirtyRegion && m_dirtyRegion->mode() == Damage::Mode::Full)
```
Let's stick to `m_dirtyRegion` as content's damage is basically the same thing. Moreover, layers can have contents and backing store at the same time so IMO there's not point in using two members for the same thing in parallel - with that we could potentially end up with more rects than needed as we'd have 2 `Damage` instances instead of one used at the same time.

I'll try to reduce the duplication and perhaps call `setNeedsDisplayInRect` as you suggested.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e8f3a51d6d77bcc828b95806628224a9cf3cfaf5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44192

nikolaszimmermann reviewed (approved):
LGTM!

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/955baa95ba394762625a26bd77df3de151cdf315. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44192

Scony committed:
[GTK][WPE] Propagate the damage from accelerated 2D canvas
https://bugs.webkit.org/show_bug.cgi?id=291678

Reviewed by Nikolas Zimmermann.

This change enables propagating damage from accelerated 2D canvas up
to the platform and to the WPE/GTK WebKit's compositor so that it can
optimize some of its drawing operations.

Co-authored-by: Carlos Garcia Campos <cgarcia@igalia.com>
Canonical link: https://commits.webkit.org/298552@main

webkit-commit-queue commented:
Committed 298552@main (fdfdd4c0e355): <https://commits.webkit.org/298552@main>

Reviewed commits have been landed. Closing PR #44192 and removing active labels.
